client/components/locale-suggestions/index.jsx

```diff
@@ -28,9 +30,19 @@ class LocaleSuggestions extends Component {
 	};
 
 	componentWillMount() {
-		if ( this.props.locale ) {
-			switchLocale( this.props.locale );
+		let { locale } = this.props;
+
+		if ( ! locale && typeof navigator === 'object' && 'languages' in navigator ) {
+			for ( const langSlug of navigator.languages ) {
+				const language = getLanguage( langSlug.toLowerCase() );
+				if ( language ) {
+					locale = language.langSlug;
+					break;
+				}
+			}
 		}
```

Comment:
we can also recognize this as n array function and if we want to can rewrite it those terms which lets us focus on the immediate step at hand and also separate the iteration from the check…

```js
if ( this.props.locale ) {
	return this.props.setLocale( this.props.locale );
}

if ( 'object' !== typeof navigator || ! navigator.hasOwnProperty( 'languages' ) ) {
	return;
}

const locale = navigator.languages.find( slug => !! getLanguage( slug.toLowerCase() ) )
if ( ! locale ) {
	return;
}

this.props.setLocale( locale.langSlug );
```

There are other changes in style I made in this example code which aren't critical for the concept.

Reply:
Thanks @dmsnell 

> we can also recognize this as n array function

Makes total sense, but in this case, I created the `for...of` iteration since we want the first valid langSlug from the config, that is, the return value of `getLanguage( langSlug.toLowerCase() )` rather than the browser's version.

The main reason, aside from ensuring that we're passing around our own slugs, is that `getLanguage` [does the following](https://github.com/Automattic/wp-calypso/blob/3c96f6f/client/lib/i18n-utils/utils.js):

1. returns a lang object if it can find it in the config, then if none is found, 
2. tries to find and return a language root if the slug is a variant, such as `de-AT`.

Oh yes, just a note on `navigator.hasOwnProperty( 'languages' )`, we'd probably have to use `! navigator.languages`, or check the positive condition of `'languages' in navigator` since we need to access `navigator`'s prototype.

Reply:
all good points. I missed that one bit about returning the transform. doesn't really change the idea much but it's not a big deal either.